chainer/distribution.py

```diff
@@ -0,0 +1,420 @@
+    `batch_shape` and `event_shape` is decided by the shape of the parameter
+    when generating an instance of a class.
+
+    .. admonition:: Example
+
+        The following code is a example of sample-batch-event hierarchy on
+        using `MultivariateNormal` distribution. This make 2d normal
+        distributions. `dist` consists of 12(4 * 3) independent 2d normal
+        distributions. And on initialization, `batch_shape` and `event_shape`
+        is decided.
+
+        >>> import chainer
+        >>> import chainer.distributions as D
+        >>> import numpy as np
+        >>> d = 2
+        >>> shape = (4, 3)
+        >>> loc = np.random.normal(
+        ...     size=shape + (d,)).astype(np.float32)
+        >>> cov = np.random.normal(size=shape + (d, d)).astype(np.float32)
+        >>> cov = np.matmul(cov, np.rollaxis(cov, -1, -2))
+        >>> l = np.linalg.cholesky(cov)
+        >>> dist = D.MultivariateNormal(loc, l)
+        >>> dist.event_shape
+        (2,)
+        >>> dist.batch_shape
+        (4, 3)
+        >>> sample = dist.sample(sample_shape=(6, 5))
+        >>> sample.shape
+        (6, 5, 4, 3, 2)
+
+    Every function about probability takes realization value whose shape is
+    `(sample_shape, batch_shape, event_shape)` and returns an evaluated value
+    whose shape is `(sample_shape, batch_shape)`.
+
+    """
+
+    def _copy_to(self, target):
+        target.__dict__ = copy.copy(self.__dict__)
+        return target
+
+    @property
+    def batch_shape(self):
+        """Returns the shape of a batch.
+
+        Returns:
+            tuple: The shape of a sample that is not identical and indipendent.
+
+        """
+        raise NotImplementedError
+
+    def cdf(self, x):
+        """Evaluates the cumulative distribution function at a given points.
+
+        Args:
+            x(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`): Data points in the domain of the
+            distribution
+
+        Returns:
+            ~chainer.Variable: Cumulative distribution function value evaluated
+            at `x`.
+
+        """
+        raise NotImplementedError
+
+    @property
+    def covariance(self):
+        """Returns the covariance of the distribution.
+
+        Returns:
+            ~chainer.Variable: The covariance of the distribution.
+        """
+        raise NotImplementedError
+
+    @property
+    def entropy(self):
+        """Returns the entropy of the distribution.
+
+        Returns:
+            ~chainer.Variable: The entropy of the distribution.
+
+        """
+        raise NotImplementedError
+
+    @property
+    def event_shape(self):
+        """Returns the shape of an event.
+
+        Returns:
+            tuple: The shape of a sample that is not identical and indipendent.
+
+        """
+        raise NotImplementedError
+
+    def icdf(self, x):
+        """Evaluates the inverse cumulative distribution function at a given points.
```

Comment:
Evaluates the inverse cumulative distribution function at a given points. ->
Evaluates the inverse cumulative distribution function at the given points.